app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt

```diff
@@ -18,78 +18,121 @@
+
+    private var lastActivityJob: Job? = null
 
-    var lastActivityLastPage = 1
     private var lastActivitiesTime: Long = 0
+    private var currentCursor: String? = null
     private var lastActivities = arrayListOf<FileActivity>()
     private var lastMergedActivities = arrayListOf<FileActivity>()
 
-    fun getLastActivities(
+    fun loadLastActivities(driveId: Int, forceDownload: Boolean = false) {
+        loadLastActivities(driveId, forceDownload, isFirstPage = true)
+    }
+
+    fun loadMoreActivities(driveId: Int) {
+        loadLastActivities(driveId, forceDownload = false, cursor = currentCursor)
+    }
+
+    fun restoreActivitiesIfNeeded() {
+        if (lastMergedActivities.isNotEmpty()) {
+            lastActivitiesResult.value = LastActivityResult(
+                mergedActivities = lastMergedActivities,
+                isComplete = currentCursor != null,
+                isFirstPage = true,
+            )
+        }
+    }
+
+    private fun loadLastActivities(
         driveId: Int,
-        forceDownload: Boolean = false
-    ): LiveData<Pair<ApiResponse<ArrayList<FileActivity>>, ArrayList<FileActivity>>?> {
-        lastActivityJob.cancel()
-        lastActivityJob = Job()
+        forceDownload: Boolean = false,
+        isFirstPage: Boolean = false,
+        cursor: String? = null,
+    ) {
+        lastActivityJob?.cancel()
+        lastActivityJob = viewModelScope.launch(Dispatchers.IO) {
+            val lastActivityResult = getLastActivities(driveId, forceDownload, isFirstPage, cursor)
+            if (lastActivityJob?.isActive == true) lastActivitiesResult.postValue(lastActivityResult)
+        }
+    }
+
+    private fun getLastActivities(
+        driveId: Int,
+        forceDownload: Boolean,
+        isFirstPage: Boolean,
+        cursor: String?,
+    ): LastActivityResult? {
 
         val ignoreDownload =
-            lastActivityPage == 1 && lastActivitiesTime != 0L && (Date().time - lastActivitiesTime) < DOWNLOAD_INTERVAL && !forceDownload
-        return liveData(Dispatchers.IO + lastActivityJob) {
-            if (ignoreDownload) {
-                lastActivityPage = lastActivityLastPage
-                emit(ApiResponse(SUCCESS, lastActivities, page = 1) to lastMergedActivities)
-                return@liveData
+            isFirstPage && lastActivitiesTime != 0L && (Date().time - lastActivitiesTime) < DOWNLOAD_INTERVAL && !forceDownload
+
+        if (ignoreDownload) {
+            return LastActivityResult(
+                mergedActivities = lastMergedActivities,
+                isComplete = false,
+                isFirstPage = true,
+            )
+        }
+
+        lastActivitiesTime = Date().time
+
+        val apiResponse = ApiRepository.getLastActivities(driveId, cursor)
+        lastActivityJob?.ensureActive()
+        val data = apiResponse.data
+
+        return if (apiResponse.isSuccess() || (isFirstPage && data != null)) {
```

Comment:
Can we have an apiResponse that is error but still having data ? it seems a bit weird